### PR TITLE
adds error_mitigation and sharpen configs

### PR DESCRIPTION
**Pull request by splch**
### Description

**Error Mitigation**
The IonQ quantum computation platform provides an error mitigation option, which can be specified in the backend configuration. To enable error mitigation, set the `error_mitigation` field in the `BackendConfig` object to a map containing `debias: true`, as follows:

```cpp
config["error_mitigation"]["debias"] = true;
```

Post Example:
```json
"error_mitigation": {
  "debias": true
}
```

**Sharpen**
The IonQ quantum computation platform provides a sharpening feature to enhance the computation precision. This feature can be controlled by setting the `sharpen` field in the `BackendConfig` object, as follows:

```cpp
config["sharpen"] = true;
```

Query Example:
```plaintext
...&sharpen=true
```

### CODE REVIEW

File: runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp [hunk 1]
```diff
@@ -197,8 +206,15 @@ std::string IonQServerHelper::constructGetResultsPath(std::string &jobId) {
...
+  // Construct the results path
+  std::string resultsPath = backendConfig.at("job_path") + jobId + "/results";
+
+  // If sharpen is true, add it to the query parameters
+  if (keyExists("sharpen") && backendConfig["sharpen"].get<bool>())
+    resultsPath += "?sharpen=true";
+
   // Return the results path
-  return backendConfig.at("job_path") + jobId + "/results";
+  return resultsPath;
```
Comment (bmhowe23) [inline runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp:217]: This code isn't called anywhere, so I suspect this should be added to the `onQServerHelper::constructGetResultsPath(ServerMessage &postResponse)` flavor of this function instead? I'll post suggestions for you to review.

**Review by bmhowe23 — COMMENTED** [review]
Thanks for these updates, @splch. Should these changes be paired with any updates to `ionq.config` to activate the changes? In any case, I'll also post a few suggested updates for my review comments below. Let me know if they meet your intentions.

File: runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp [hunk 2]
```diff
@@ -92,6 +92,11 @@ void IonQServerHelper::initialize(BackendConfig config) {
   // Construct the API job path
   backendConfig["job_path"] =
       backendConfig["url"] + '/' + backendConfig["version"] + "/jobs";
+  // Enable error mitigation
+  if (config.find("error_mitigation") != config.end())
+    backendConfig["error_mitigation"] = config["error_mitigation"];
```
[the remark this answers is not in the document]
Comment (splch) [reply runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp]: got it thanks!

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config [hunk 3]
```diff
@@ -49,6 +51,14 @@ while [ $# -gt 1 ]; do
 		PLATFORM_EXTRA_ARGS="$PLATFORM_EXTRA_ARGS;noise;$2"
 		IONQ_NOISE_MODEL="$2"
 		;;
+	--error-mitigation)
```
Comment (bmhowe23) [inline runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config:54]: Per the comment in lines 40-41, this command line option needs to start with `ionq`
```suggestion
	--ionq-error-mitigation)
```

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config [hunk 4]
```diff
@@ -49,6 +51,14 @@ while [ $# -gt 1 ]; do
...
+		PLATFORM_EXTRA_ARGS="$PLATFORM_EXTRA_ARGS;error_mitigation;$2"
+		ERROR_MITIGATION="$2"
+		;;
+	--sharpen)
```
Comment (bmhowe23) [inline runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config:58]: Per the comment in lines 40-41, this command line option needs to start with `ionq`.
```suggestion
	--ionq-sharpen)
```

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config (hunk 3 above)
Comment (splch) [reply runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config:54]: ah i see! thanks

File: runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp (hunk 1 above)
Comment (splch) [reply runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp:278]: True - though I'd like to include sharpen as an option if someone needs to get a job by id. Is it fine to keep sharpen in both methods?

File: runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (splch) [reply runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp]: alright i updated the code so we can set `debias=true` which will later create the `error_mitigation={"debias": true}` for the job.

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config [hunk 5]
```diff
@@ -33,6 +33,8 @@ LIBRARY_MODE=false
 # Default ionq-machine target if not specified.
 IONQ_MACHINE="simulator"
 IONQ_NOISE_MODEL=""
+debias=false
+SHARPEN=false
```
Comment (bmhowe23) [inline runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config:37]: These two shell variables aren't used anywhere. Should they be deleted?
```suggestion
```

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config [hunk 6]
```diff
@@ -49,6 +51,14 @@ while [ $# -gt 1 ]; do
 		PLATFORM_EXTRA_ARGS="$PLATFORM_EXTRA_ARGS;noise;$2"
 		IONQ_NOISE_MODEL="$2"
 		;;
+	--ionq-debias)
+		PLATFORM_EXTRA_ARGS="$PLATFORM_EXTRA_ARGS;debias;$2"
+		DEBIAS="$2"
```
Comment (bmhowe23) [inline runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config:56]:
```suggestion
```

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config [hunk 7]
```diff
@@ -49,6 +51,14 @@ while [ $# -gt 1 ]; do
...
+		;;
+	--ionq-sharpen)
+		PLATFORM_EXTRA_ARGS="$PLATFORM_EXTRA_ARGS;sharpen;$2"
+		SHARPEN="$2"
```
Comment (bmhowe23) [inline runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config:60]:
```suggestion
```

File: runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp [hunk 8]
```diff
@@ -152,19 +193,33 @@ IonQServerHelper::createJob(std::vector<KernelExecution> &circuitCodes) {
...
       job["noise"] = noiseModel;
     }
 
     job["qubits"] = backendConfig.at("qubits");
-    job["shots"] = static_cast<int>(shots);
+    job["shots"] = shots;
     job["input"]["format"] = "qir";
     job["input"]["data"] = circuitCode.code;
+    // Include error mitigation configuration if set in backendConfig
+    // Include error mitigation configuration if set in backendConfig
```
Comment (bmhowe23) [inline runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp:211]:
```suggestion
    // Include error mitigation configuration if set in backendConfig
```

**Review by bmhowe23 — COMMENTED** [review]
It would be nice to also update the IonQ section of the documentation (in `docs/sphinx/using/hardware.rst`) to explain how/when/why the user should use `--ionq-debias` and `--ionq-sharpen`. If that's too much trouble, then you could put it in this comment and I could update it after this PR gets merged.

**Comment by bmhowe23** [comment]
I merged from main and applied updates based on my review comments. (Hopefully that's ok...we've got code cutoff for 0.5.0 coming very quickly now, and I figured it would be best to push this along.)

I also ran all the IonQ tests that the nightly integration test would run, and they passed.

File: runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp (hunk 8 above)
Comment (bmhowe23) [reply runtime/cudaq/platform/default/rest/helpers/ionq/IonQServerHelper.cpp]: Updated in 56051c7a

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config (hunk 7 above)
Comment (bmhowe23) [reply runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config]: Updated in 56051c7a

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config (hunk 6 above)
Comment (bmhowe23) [reply runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config]: Updated in 56051c7a

File: runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config (hunk 5 above)
Comment (bmhowe23) [reply runtime/cudaq/platform/default/rest/helpers/ionq/ionq.config]: Updated in 56051c7a

**Review by bmhowe23 — APPROVED** [review]

**Pull request merged**